### ISSUE TITLE
VoiceOver [iOS] no longer respects scopesRoute and namesRoute in Flutter 1.22

### ISSUE THREAD

**Issue by dannyvalentesonos** [flutter/flutter#67345]
In Flutter 1.19, the behavior with the code below is that VoiceOver would successfully read "Page 1 -> Hello, press the button below to launch the next page".

Selecting and double tapping on "Launch route 2" would result in VoiceOver reading "Page 2 -> Hello, press the button below to close this page".

Going back and forth between page routes would work nicely, and the first element on the page would always be selected and read.

However, in Flutter 1.22, that's no longer the case. The semantics label is random whether it's read, and the first element on the page is also random to whether it gets selected or not.

This breaks our experience for visually impaired uses in our app.

```dart

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: SemanticsNav1HomePage(),
    );
  }
}

class SemanticsNav1HomePage extends StatelessWidget {

  @override
  Widget build(BuildContext context) {

    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.dark);

    return Semantics(
      scopesRoute: true,
      namesRoute: true,
      label: "Page 1",
      explicitChildNodes: true,
      child: Scaffold(
        body: SafeArea(
          top: true,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            mainAxisSize: MainAxisSize.max,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Padding(
                padding: const EdgeInsets.all(24.0),
                child: Text(
                  "Hello, press the button below to launch the next page",
                  style: TextStyle(fontSize: 24),
                ),
              ),
              Expanded(
                child: Center(
                  child: MaterialButton(
                      color: Colors.black,
                      textColor: Colors.white,
                      child: Text("Launch route 2", style: TextStyle(fontSize: 20)),
                      onPressed: () => Navigator.push(context, PageRouteBuilder(pageBuilder: (context, animation, secondaryAnimation) => SemanticsNav2HomePage())),
                  )
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class SemanticsNav2HomePage extends StatelessWidget {

  @override
  Widget build(BuildContext context) {

    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.dark);

    return Semantics(
      scopesRoute: true,
      namesRoute: true,
      label: "Page 2",
      explicitChildNodes: true,
      child: Scaffold(
        body: SafeArea(
          top: true,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            mainAxisSize: MainAxisSize.max,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Padding(
                padding: const EdgeInsets.all(24.0),
                child: Text(
                  "Hello, press the button below to close this page",
                  style: TextStyle(fontSize: 24),
                ),
              ),
              Expanded(
                child: Center(
                    child: MaterialButton(
                      color: Colors.black,
                      textColor: Colors.white,
                      child: Text("Pop route 2", style: TextStyle(fontSize: 20)),
                        onPressed: () => Navigator.pop(context),
                    )
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```

<details>
<summary>Flutter doctor</summary>

[✓] Flutter (Channel beta, 1.22.0-12.2.pre, on Mac OS X 10.15.6 19G2021, locale en-US)
• Flutter version 1.22.0-12.2.pre at /Users/danny.valente/Library/flutter/1.22.0-12.2.pre-beta
• Framework revision 2bafdc8 (10 days ago), 2020-09-25 12:48:22 -0700
• Engine revision f763b5b9b9
• Dart version 2.10.0 (build 2.10.0-110.5.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
• Android SDK at /Users/danny.valente/Library/Android/sdk
• Platform android-29, build-tools 29.0.2
• Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
• Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
• All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
• Xcode at /Applications/Xcode.app/Contents/Developer
• Xcode 11.4.1, Build version 11E503a
• CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
• Android Studio at /Applications/Android Studio.app/Contents
• Flutter plugin version 47.1.2
• Dart plugin version 193.7361
• Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

</details>

**Comment by dannyvalentesonos** [comment]
I'm attaching two videos. One with Flutter 1.19 (good) and the other with 1.22 (bad).

[vo_issues.zip](https://github.com/flutter/flutter/files/5329977/vo_issues.zip)

**Comment by dannyvalentesonos** [comment]
@xster Another regression in 1.22 Thanks!

**Comment by dannyvalentesonos** [comment]
Just for some context, we are using Add2App, and we have a flutter view that can pop at any point to help the user do a particular needed action at a given time, and now when that pops up, the namesRoute label sometimes is read, sometimes it's not read at all. And in all cases, nothing in the flutter view is actually selected by VO.

So this is pretty broken for us when VO is on.

**Comment by xster** [comment]
Is this Android and iOS or iOS only?

**Comment by xster** [comment]
@gaaclarke could I ask you to take a look at this one?

**Comment by dannyvalentesonos** [comment]
> Is this Android and iOS or iOS only?

iOS only.

**Comment by chunhtai** [comment]
Maybe a regression of https://github.com/flutter/engine/pull/19826/files

**Comment by gaaclarke** [comment]
Yep, I was able to verify that 48a6fde39f9fe8bf7ad06f0f8cf690186d930b73 introduced the regression.

I can reproduce this bug, however the behavior is deterministic for me.  It starts to say the page name then interrupts itself to start describing the next focused item.  On the first page it is the text, on the second page it is the button.

**Comment by dannyvalentesonos** [comment]
@gaaclarke in our Add2App scenario, it's not as deterministic at all, but might be the result of the same issue you describe. It seems that whatever VoiceOver reads depends on what was selected before the Flutter view launches. Sometimes it will read the status bar, and other times it won't even read the page name or anything else at all, thereby not even letting the user know there's a new view in front of them.

**Comment by gaaclarke** [comment]
I put a print statement in the engine here:
```diff
--- a/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
+++ b/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
@@ -31,6 +31,7 @@ class DefaultIosDelegate : public AccessibilityBridge::IosDelegate {
 
   void PostAccessibilityNotification(UIAccessibilityNotifications notification,
                                      id argument) override {
+    NSLog(@"gaaclarke %@", argument);
     UIAccessibilityPostNotification(notification, argument);
   }
 };
```

With 48a6fde:
```
2020-10-06 16:54:26.044254-0700 Runner[1149:534279] gaaclarke FlutterSemanticsObject [0x108cad8b0] Page 1 {{0, 0}, {375, 667}}
2020-10-06 16:54:34.775664-0700 Runner[1149:534279] gaaclarke (null)
2020-10-06 16:54:34.804867-0700 Runner[1149:534279] gaaclarke FlutterSemanticsObject [0x108bb55b0] Page 2 {{0, 0}, {375, 667}}
2020-10-06 16:54:44.557695-0700 Runner[1149:534279] gaaclarke FlutterSemanticsObject [0x108b4ab90] Page 1 {{0, 0}, {375, 667}}
2020-10-06 16:54:44.865602-0700 Runner[1149:534279] gaaclarke FlutterSemanticsObject [0x108b4a9a0] Hello, press the button below to launch the next page {{24, 44}, {327, 87}}
```

Without 48a6fde:
```
2020-10-06 17:02:45.605097-0700 Runner[1157:537018] gaaclarke Page 1
2020-10-06 17:03:06.935834-0700 Runner[1157:537018] gaaclarke (null)
2020-10-06 17:03:06.968285-0700 Runner[1157:537018] gaaclarke Page 2
2020-10-06 17:03:19.258915-0700 Runner[1157:537018] gaaclarke Page 1
2020-10-06 17:03:19.545795-0700 Runner[1157:537018] gaaclarke FlutterSemanticsObject [0x10a8d8440] Hello, press the button below to launch the next page {{24, 44}, {327, 87}}
```

I guess that's expected.  It's just surprising that the change of the datatype causes such a different behavior.  @chunhtai how did this PR fix your issue?

**Comment by chunhtai** [comment]
The original issue is that it will announce the string and then focus the first item on the screen. If the first item is the namesRoute, it will get pronounce twice.

Changing the data type from string to object will make it focus the object directly.

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.23.0-8.0.pre.222`.
<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.23.0-8.0.pre.222, on Mac OS X 10.15.7 19H2 x86_64, locale en-GB)
    • Flutter version 1.23.0-8.0.pre.222 at /Users/markus/development/flutter_master
    • Framework revision d2858f00dd (3 hours ago), 2020-10-08 09:39:13 +0200
    • Engine revision 2eac514f26
    • Dart version 2.11.0 (build 2.11.0-186.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/markus/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (5 available)
    • Nevercode’s iPhone (mobile)         • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.7
    • iPhone SE (2nd generation) (mobile) • B46F1F22-93B0-40C4-8DA7-1C9816E399F7     • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • macOS (desktop)                     • macos                                    • darwin-x64     • Mac OS X 10.15.7 19H2 x86_64
    • Web Server (web)                    • web-server                               • web-javascript • Flutter Tools
    • Chrome (web)                        • chrome                                   • web-javascript • Google Chrome 85.0.4183.121

• No issues found!
```
</details>

**Comment by gaaclarke** [comment]
Okay, I propose we revert https://github.com/flutter/engine/pull/19826.  I looked into the issue that that PR fixed (https://github.com/flutter/flutter/issues/44001) and that just looks like the reported error is working as intended.  When the route changes, it is announced and the first item is selected.  If that first item happens to have the same name as the route then the same words will get said twice, once for the route name, once for the focused item.

We don't have any control over the reading of the focused item, the only thing we could do is not announce the route name which would be erroneous in other cases.  We don't know with certainty what VoiceOver will focus on so we can't preemptively avoid saying the route name if it is the same as the focused item.  I played around with calling `UIAccessibilityFocusedElement` to do that kind of logic but its asynchronous nature and limitations of the API make it not possible.

If we really want to get rid of the words getting being said twice we should look into removing the route name of the AlertDialog whose first item has the same semantic label as the route name.

Sound good, @chunhtai ?

**Comment by chunhtai** [comment]
@gaaclarke yeah, that sounds good to me.

**Comment by gaaclarke** [comment]
closed via https://github.com/flutter/engine/pull/21714

**Comment by dannyvalentesonos** [comment]
I see that a new flutter 1.22 stable release was just released 3 days ago, but this bug still does not have the "cp: 1.22 completed" label. When can this be expected to get cherry picked to 1.22?

Thanks!
